assets/helpers/paymentIntegrations/stripeCheckout.js

```diff
@@ -1,13 +1,155 @@
 // @flow
 
+/**
+ * This module contains a copy of pages/oneoff-contributions/helpers/ajax so that
+ * the new payment flow can use this code without any impact on the existing
+ * infrastructure.
+ *
+ * There are a couple of differences though:
+ * - the amount: instead of fixing it when the callback is created, it should
+ *   be provided by the `getState` function
+ * - the actions: what happens upon error/success is view-dependent, so the
+ *   actions are provided as parameters too. As a result, `dispatch` can
+ *   go away (which makes the module completely independant of React)
+ * - tracking conversions: this should be decoupled from the payment itself and
+ *   moved upstream, in the success handler
+ *
+ * The latter module can be removed entirely once the new payment flow becomes the one
+ * and only contribution endpoint.
+ */
+
+
+import { addQueryParamsToURL } from 'helpers/url';
+import {
+  derivePaymentApiAcquisitionData,
+  type ReferrerAcquisitionData,
+  type PaymentAPIAcquisitionData,
+} from 'helpers/tracking/acquisitions';
+import type { Participations } from 'helpers/abTests/abtest';
+import type { IsoCurrency } from 'helpers/internationalisation/currency';
+import type { OptimizeExperiments } from 'helpers/tracking/optimize';
+
+import * as cookie from 'helpers/cookie';
+import { logException } from 'helpers/logger';
 
 // ----- Setup ----- //
 
 let stripeHandler = null;
 
+const ONEOFF_CONTRIB_ENDPOINT = window.guardian.paymentApiStripeEndpoint;
+
+function stripeOneOffContributionEndpoint(testUser: ?string) {
+  if (testUser) {
+    return addQueryParamsToURL(
+      ONEOFF_CONTRIB_ENDPOINT,
+      { mode: 'test' },
+    );
+  }
+
+  return ONEOFF_CONTRIB_ENDPOINT;
+}
+
+// ----- Types ----- //
+
+type PaymentApiStripeExecutePaymentBody = {|
+  paymentData: {
+    currency: IsoCurrency,
+    amount: number,
+    token: string,
+    email: string
+  },
+  acquisitionData: PaymentAPIAcquisitionData,
+|};
+
+type CheckoutData = {|
+  user: {
+    fullName: string,
+    email: string,
+  },
+  amount: number
+|};
 
 // ----- Functions ----- //
 
+function requestData(
+  abParticipations: Participations,
+  paymentToken: string,
+  currency: IsoCurrency,
+  referrerAcquisitionData: ReferrerAcquisitionData,
+  optimizeExperiments: OptimizeExperiments,
+  getData: () => CheckoutData,
+) {
+  const { user, amount } = getData();
+
+  const oneOffContribFields: PaymentApiStripeExecutePaymentBody = {
+    paymentData: {
+      currency,
+      amount,
+      token: paymentToken,
+      email: user.email,
+    },
+    acquisitionData: derivePaymentApiAcquisitionData(referrerAcquisitionData, abParticipations, optimizeExperiments),
+  };
+
+  return {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify(oneOffContribFields),
+    credentials: 'include',
+  };
+}
+
+function postToEndpoint(request: Object, onSuccess: Function, onError: Function): Promise<*> {
```

Comment:
Could we make use of the fact that this returns a promise rather than passing in callbacks? We could `throw` on the error conditions and the caller could supply a `.catch()`

Reply:
Hmm, I don't think so. This logic is happening _after_ stripe checkout, it is invoked via the supplied `token` callback in `Stripe.configure` ([here](https://github.com/guardian/support-frontend/blob/8f41da5a95dad01da0218456a8abb0b7eca714cd/assets/helpers/paymentIntegrations/stripeCheckout.js#L213))